File: haskell/perfect-numbers/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification (..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify x | x <= 0 = Nothing
classify x
  | aliqotSum == x = Just Perfect
  | aliqotSum > x = Just Abundant
  | otherwise = Just Deficient
  where
    aliqotSum = sum [y | y <- [1 .. (x `div` 2)], x `mod` y == 0]

File: typescript/perfect-numbers/perfect-numbers.ts
export const classify = (n: number): string => {
  if (n <= 0) throw new Error('Classification is only possible for natural numbers.')

  const sum = factorSum(n)

  if (sum === n) return 'perfect'
  if (sum > n) return 'abundant'

  return 'deficient'
}

const factorSum = (num: number): number => {
  return [...Array(Math.floor(num))]
    .reduce((acc: number, _: number, i: number) =>
      num % i === 0
        ? acc + i
        : acc, 0)
}
